Adding: Types, Tests, WebUSB, Web Example and more

gozmanyoni opened the pull request:
This PR Covers:

- Migrated code to TS to introduce types - less error prone and more robust
- Refactored the code, separated concerns a bit and moved message parsing into messages instead of DongleHandler/Parsing
- Migrated to WebUsb - which works both in node and browser - this can now run in chromium independently  with no websocket server (Also added a working example)
- From testing - it seems that the asset downloading/copying is actually not needed - the dongle works without this step.
- Introduced entry points in the package, so it can be built for both node and web, and consumed in both through these
- Added tests for DongleDriver and messages with ~90% coverage
- Linting/Formatting/EditorConfig

gozmanyoni commented on src/modules/messages.ts:
```diff
@@ -0,0 +1,544 @@
+
+export class HeaderBuildError extends Error {}
+
+export class MessageHeader {
```
@rhysmorgan134 I moved the logic of parsing/constructing messages here - should be easier to write tests for individual messages that way

rhysmorgan134 commented:
Wow, thanks so much Yoni, I will have to take some real time to look through it, but looks like I can learn a bunch from this!

gozmanyoni commented on examples/carplay-web-app/src/worker/carplay.ts:
Using a webworker here to offload decoding and emitting messages from the main thread

rhysmorgan134 commented:
Excited to give this a go, have you noticed much improvement with the web worker?

gozmanyoni commented:
@rhysmorgan134 I pushed an update with tests, and 2 examples with both a webworker one and direct consumption.
On a Pi 4 the direct example is snappy with 60fps and smooth audio - worth seeing if a Pi 3 will benefit with a webworker as that should clear up the main thread a bit.

Otherwise on a modern Pi the worker seems like an overkill

rhysmorgan134 commented on examples/carplay-web-app-worker/package.json:
```diff
@@ -0,0 +1,54 @@
+    "react-scripts": "5.0.1",
+    "stream-browserify": "^3.0.0",
+    "typescript": "^4.9.5",
+    "web-vitals": "^2.1.4"
```
@gozmanyoni Only issue I ran into when testing was a type declaration error for web-vitals inside the examples, I bumped the version up to 3.4.0 and that seemed to resolve it

gozmanyoni replied:
Bumped the version and pushed :) 

Glad to hear its working well for you too! 🥳 
As for TS - I thought it will be useful to use here - as the library is quite big, lots of messages, enums and so on - it can become very fragile over time and TS really help in these cases - types are really informative and help when dealing with external lib's or browser api's as well - If you are happy to merge this at any point we can also add github actions that for every branch/PR will verify that:

- Linting rule check passes
- Type Check passes
- Tests Pass

steelbrain commented:
@gozmanyoni is there a way I could buy you some coffee?

gozmanyoni commented:
Thanks @steelbrain - Never had a page but I set one up :) https://www.buymeacoffee.com/ygoz I do love my morning coffees

steelbrain commented:
Ran this on a Raspberry pi 3 model b, here's the experience so far:
- Had to install `libudev-dev` to successfully compile node-usb
- Direct worked but was a little choppy (not complaining, still pretty good!)
- Worker was smoother

I could be biased, objective measurement would prove more reliable

gozmanyoni commented:
Thanks for testing @steelbrain, I will check the touch/drag issues out - must be a bug in the example, haven't tested touch myself yet.

Will be improving test coverage as well - still need to test a bit more of the Driver, and all SendableMessage classes

gozmanyoni commented:
@steelbrain Fixed touch - apparently browsers will send a pointercancel soon after they start handling touch, we can specify there needs to be no touch-action handled by the browser to prevent this - as I did https://developer.mozilla.org/en-US/docs/Web/CSS/touch-action.

Its ok for this use-case as we don't want the browser to do anything on the video element or div containing it, touch is handled internally.

steelbrain commented:
Confirmed, touch and swipe is working flawlessly. Thank you 🚀

rhysmorgan134 commented on src/node/CarplayWS.ts:
```diff
@@ -0,0 +1,104 @@
+  DEFAULT_CONFIG,
+} from '../modules'
+
+export default class CarplayWS extends EventEmitter{
+  private _pairTimeout: NodeJS.Timeout | null = null
+  private _plugged = false
+  private _dongleDriver: DongleDriver
+
+  constructor(config: DongleConfig = DEFAULT_CONFIG, port = 5005) {
+    super()
```
Extends EventEmitter

rhysmorgan134 commented on src/node/CarplayWS.ts:
```diff
@@ -0,0 +1,104 @@
+        }
+        this._plugged = true
+        this.emitPlugged()
+      } else if (message instanceof Unplugged) {
+        this._plugged = false
+        this.emit('quit')
+      } else if (message instanceof VideoData) {
+        this.emit('carplay', message)
+      } else if (message instanceof AudioData) {
+        this.emit('audio', message)
+      } else if (message instanceof MediaData) {
+        this.emit('media', message)
+      }
```
These all now emit the message from CarplayWS

rhysmorgan134 replied:
```JavaScript
  carplay.on('carplay', (message) => {
    mainWindow?.webContents.send('videoData', message);
  });

  carplay.on('audio', (message) => {
    mainWindow?.webContents.send('audioData', message);
  });
```

Example of how I have currently implemented it for electron

gozmanyoni replied:
This looks great!

You might want to also emit carplay messages like CarplayWeb does:

`else if (message instanceof CarPlay) {
        this.onmessage?.({ type: 'carplay', message })
      }`
      
     That way - you can get a message if users click the Car icon in the carplay interface and switch to you own react app

rhysmorgan134 commented on src/node/CarplayWS.ts:
```diff
@@ -0,0 +1,104 @@
+    this._dongleDriver = driver
+  }
+
+  getStatus = () => {
+    this.emitPlugged()
+  }
+
+  sendTouch = ({ type, x, y }: { type: number; x: number; y: number }) => {
+    this._dongleDriver.send(new SendTouch(x, y, type))
+  }
```
Two new external functions, get status, and send touch

rhysmorgan134 commented on src/node/CarplayWS.ts:
```diff
@@ -0,0 +1,104 @@
+    }
+  }
+
+  private emitPlugged = () => {
+    this.emit('status', {
+      status: this._plugged,
+    })
+  }
```
This also emits direct now from CarplayWS

rhysmorgan134 reviewed (commented):
Changes highlighted below

rhysmorgan134 commented:
I'm in the process of creating a MOST bus version of carplay using this PR, so far audio is going well and fully working, I'm also transferring all other messages over IPC seems to work great too.

rhysmorgan134 commented:
> Just documenting a tangent from electron, instead of having to package an electron app, the whole carplay experience can now be shared around as static output for people to use. For example, I'm using it in my project car for carplay with off-the-shelf `chromium-browser` & `sway`. I've added the following line to my sway config, and setup [pm2](https://www.npmjs.com/package/pm2) to autostart the static file server and it's a very seamless experience, you boot the thing up and you get Apple Carplay! Thank you both, this is so cool
> 
> ```
> exec /usr/bin/chromium-browser --enable-experimental-web-platform-features --noerrdialogs --disable-infobars --disable-webusb-security --kiosk http://localhost:3000
> ```
> 
> Re MOST bus for audio, it would be interesting to explore the idea of having a simple websocket server (maybe the same one that serves these static files) that accepts audio and is used for IPC. Not sure whether it'll be more elegant than electron packaged into a single app image, but it would make distribution heck of a lot easier

Interesting concept, to be honest it has never crossed my mind to do it your method, but it seems far simpler than maintaining an electron build, I would guess likely more performant too, there's various issues with Electron and Raspberry Pi around hardware accelleration!

Also thanks alot for the coffees, really appreciated!

gozmanyoni commented:
@rhysmorgan134 Changes look great - One nitpick about naming - perhaps we should rename CarPlayWS now - as it no longer uses WebSockets, rather CarPlayManager or something. 

It basically serves as an orchestration layer hooking the driver up with the mic and handling init logic and subscriptions.

Curious if you had a chance to play with the Microphone over MOST - I will try to interact with the mic in my car over the weekend

rhysmorgan134 commented:
> @rhysmorgan134 Changes look great - One nitpick about naming - perhaps we should rename CarPlayWS now - as it no longer uses WebSockets, rather CarPlayManager or something.
> 
> It basically serves as an orchestration layer hooking the driver up with the mic and handling init logic and subscriptions.
> 
> Curious if you had a chance to play with the Microphone over MOST - I will try to interact with the mic in my car over the weekend

Ha I did think of that whilst doing the change, I think I need to remove Socket.io from dependencies too, I'll do this and add the carplay message aswell. I've been setting up an electron version using the node build. Audio and video are working, I have a small bug with the video which seems to be that first message, which would be a keyframe must be getting missed so there's a scrambled feed until the next key frame, high likelyhood it's my implementation unless you experience this too? It's been a great learning experience for typescript as sticking with it for electron too!

I haven't had a chance to play around with the microphone however the flow should be:

- Send an allocate message (0x101) to the microphone with an unsigned byte for the source number (will likely be 1), with an OpType of StartResult (0x02 the same as SetGet, I haven't implemented the methods OpTypes yet to most explorer, check here for more info https://github.com/rhysmorgan134/most-explorer/blob/main/src/main/utils/enums.js)
- Response should be
  - unsigned byte - sourceNr
  - unsigned byte - srcDelay
  - stream of unsigned bytes for each channel
- Those bytes will then need to be routed to the correct position in the routing table to output to sx1, _in theory this should be locations 46 56, 66, 76_ this will need a similar function to the one used to route data onto the network which is here https://github.com/rhysmorgan134/SocketMost/blob/7b6698429dca8c8add3ec025a2abbf922b81336a/OS8104A.js#L400C1-L423C6

In theory all audio then found on those channels should appear as audio input to the pi, I suspect it will be two bytes of audio as I guess it's going to be mono rather than stereo. 

I'm trying to get hold of a microphone to use on my test set up which I'd be happy to do the experimenting on with that, as it's easier than doing it on actual car!

steelbrain commented:
Noting additional information based on experience so far
- On Raspberry pi 4 model b, worker still feels smoother than direct (especially with audio playing, gpu accel turned off)
- In Kiosk mode, chromium-browser doesn't allow audio autoplay, so have to add `--autoplay-policy=no-user-gesture-required` to make it play Music etc
- By default `chromium-browser` uses GPU acceleration, which is slower for whatever reason, so you have to turn it off for better performance (`--disable-gpu`)
- Audio mileage may vary, for me I had to install `libasound2-plugins` and configure `~/.asoundrc` to be the following
```
pcm.!default {
  type plug
  slave {
    pcm "plughw:3,0"
  }
}

ctl.!default {
  type hw
  card 3
}
```
You can find your card number for above through `arecord -l`

------
Video is pretty sweet right now but it's pretty hard to listen to music, audio is choppy/drops off (both worker and direct, tho worker is better for audio). I have limited knowledge in this area, but **is there any opportunity for audio optimization we're leaving on the table**?

I noticed that the CPU utilization shoots up when music is playing. I will try some profiling later and post if I find anything interesting

rhysmorgan134 commented:
@steelbrain are you running with a physical screen plugged in or are you running with a headless desktop with vnc? Also what resolution and frame rate? Would be good to see if I experience the same using the node build but I would want to match your set up for a direct comparison

steelbrain commented:
I am using Raspberry Pi 4 model 4GB version with [Waveshare 7 inch touchscreen](https://botland.store/raspberry-pi-displays/4688-touch-screen-c-capacitive-lcd-7-1024x600px-hdmi-usb-for-raspberry-pi-32bzero-waveshare-11199-5904422333010.html) running 1024x600 resolution over HDMI. With GPU accel turned off in chrome and raspi video memory set to 256mb

gozmanyoni commented:
@steelbrain hmm, I tested it on a Pi 4 model 8gb, with a 64bit OS and a large HDMI screen (1440p), although made sure the browser window aligns with the resolution you have - audio and video are playing smoothly, CPU at 30-60%, I think worker has slightly lower CPU usage and less spikes on cpu usage - but overall both work well.

Chromium is on default settings which I believe enable GPU accel - I do not have a usb soundcard though, audio goes over HDMI on the bench, over PiMOST in the car (Which i have not tested with carplay yet, just audio from the browser)

rhysmorgan134 commented:
Is it worth testing this to see if you experience the same?

https://github.com/rhysmorgan134/react-carplay/releases/tag/v3.0.2

@gozmanyoni I have gotten microphone working this morning, althought I have done it by retrieving the cars cd player stream and writing to a file, that works well, a bit unsure how to handle the two byte mono stream, whether to send as just one channel, or adjust the overlay to make it 48khz single channel

gozmanyoni commented:
@rhysmorgan134 oh wow! quick progress! 🥳 
so its a two byte 48khz mono pcm stream? with stereo audio its just two bytes left channel and two bytes right as far as i know

rhysmorgan134 commented:
yeah correct, so when I do an allocate to the Microphone Function Block, it returns two bytes, I am guessing it has to be the same sample rate as the most bus so 48khz. Normal streams are 4bytes so 2 for left 2 for right, I might see if I can just send the two bytes to both left and right outputs, so that It's technically stereo. I've tested by feeding the 4 byte audio stream the cd player into the pi as the microphone feed and works well. I don't really want to make the overlay input a mono input as then we would lose the ability of streaming stereo into the pi (if that's ever needed)

gozmanyoni commented:
@rhysmorgan134 Yeah i think its good that you maintain a stereo format from the lowest level possible in the implementation, if a vehicle has a stereo microphone it will be easy to adjust.

Update on the DongleDriver - I looked around the APK and found out how we can also set the dongles mediaDelay prop - this can reduce the delay in playing audio. Implemented and added to the box initialization logic and examples.

steelbrain commented:
Update re audio quality, after plugging in my dongle to an Android device, I've concluded that the bad audio is because of the dongle, so no need to dig deeper into that. I'll try to get a replacement. Thanks y'all

rhysmorgan134 commented:
> @rhysmorgan134 Yeah i think its good that you maintain a stereo format from the lowest level possible in the implementation, if a vehicle has a stereo microphone it will be easy to adjust.
> 
> Update on the DongleDriver - I looked around the APK and found out how we can also set the dongles mediaDelay prop - this can reduce the delay in playing audio. Implemented and added to the box initialization logic and examples.

Very nice, I had been looking for that one for a long time but never found it in my APK endeavours! I think that same message may give the ability to enable the media info to?

This has made me think of something which I may try. When I was first developing node-carplay/react-carplay all attempts to play video would fail after a certain amount of time taking up all the RAM. It turned out the dongle only sent a keyframe on start up and then on request, I managed to find the keyframe request in the APK, and I sent it at a fairy frequent rate, in my jubilation I never went back and upped it so I believe its still 5 seconds. I'm not sure whether we should up this to something much higher as I suspect it may improve performance on larger resolution displays.

gozmanyoni commented:
@rhysmorgan134 I upped it to 5 - i think it used to be an interval of 1 sec - and I played with it bit as well - but yeah we can keep changing it and find out where it in fact does start to introduce a delay in the video!

What do you mean by enabling media info? is it the media messages that we are parsing?
It seems that this method send a json object to the dongle, but it only has the mediaDelay, a timestamp, and then a few android auto settings.

The APK seems to have the ability to handle different microphone sample rates, and sent a microphone type to the dongle - I wonder if we can configure that too.

I noticed that my web example has a bug with calls, as the time passes the microphone input becomes delayed, looking into that

rhysmorgan134 commented:
I think on a 2gb pi it lasted around 5minutes before filling up, if I understand it correct jmuxer holds all the previous h264 data until it receives a key frame, so I think we should be good for a larger delay.

For my dongle before I could receive media data I had to go onto the dongles webpage and enable “car integration” (I may have remembered the name wrong) but it’s on the same settings page as the audio delay.

gozmanyoni commented:
Pushed another update to this, figured out more things from the apk:

1. Can now set a custom label and icon for the home button
2. can configure what home button does (it can either toggle Siri or emit an event)

Added a (slightly hacky) script to configure the icon and label - but we can also build a web interface to allow this.

gozmanyoni commented:
Pushed an update

* Microphone now using AudioWorkletProcessor - the new, more performance and not deprecated API - Sadly, it still develops a delay during a call that is quite noticeable after ~30 seconds.
* Worker example actually had a bug - workers will always deep clone any data posted from them or to them - with audio and video streams that's significant, but it can be prevented by using Transferable Objects - https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Transferable_objects, updated the example to utilize this.

steelbrain commented:
Built and ran the latest commit, here's what I get. Audio worklet fails to load because it has a bunch of require statements.


<img src="https://github.com/rhysmorgan134/node-CarPlay/assets/4278113/2b9db666-d049-40ee-ab22-1e9a77c23595">

<img src="https://github.com/rhysmorgan134/node-CarPlay/assets/4278113/01e8b888-643a-4a23-9724-e8e9e8d7583b">

steelbrain commented:
New dongles arrived, here's a recording of the audio artifacts that I am experiencing (both macos/raspi, tried at least 3 carplay dongles). It's the clicking/popping sound, does not happen when the APK is used (tried on lineageos 20 on the raspi) but does happen when using node-carplay

[bug-demo.webm](https://github.com/rhysmorgan134/node-CarPlay/assets/4278113/f4c88a64-b3af-4b0c-99f1-5c5d7c45b86a)

gozmanyoni commented:
Thanks @steelbrain - First issue is due to the format I am importing the audio worklet processors in the library - it works well with the react dev-server but will need more tweaking to work well in a production build - will address this over the weekend.

For audio, I noticed this myself and have been working on a more optimised audio playback solution.
I am curious - do you experience that with https://github.com/rhysmorgan134/react-carplay/releases/tag/v3.0.2 as well? or is it a degradation form my branch?

Lastly - i will be merging this PR into a new "next" branch in this repo, and will follow up with smaller PR's for fixes and tweaks until its stable enough to be released and merged to master

steelbrain commented:
Re audio worklet, here's the patch I use

<details>

<summary>Patch</summary>

```patch
diff --git a/examples/carplay-web-app-direct/package.json b/examples/carplay-web-app-direct/package.json
index 526829b..e896631 100644
--- a/examples/carplay-web-app-direct/package.json
+++ b/examples/carplay-web-app-direct/package.json
@@ -42,9 +42,9 @@
   },
   "browserslist": {
     "production": [
-      ">0.2%",
-      "not dead",
-      "not op_mini all"
+      "last 1 chrome version",
+      "last 1 firefox version",
+      "last 1 safari version"
     ],
     "development": [
       "last 1 chrome version",
diff --git a/examples/carplay-web-app-worker/package.json b/examples/carplay-web-app-worker/package.json
index 2911e03..4593521 100644
--- a/examples/carplay-web-app-worker/package.json
+++ b/examples/carplay-web-app-worker/package.json
@@ -42,9 +42,9 @@
   },
   "browserslist": {
     "production": [
-      ">0.2%",
-      "not dead",
-      "not op_mini all"
+      "last 1 chrome version",
+      "last 1 firefox version",
+      "last 1 safari version"
     ],
     "development": [
       "last 1 chrome version",
```

</details>

steelbrain commented:
Re `react-carplay`, I've never had any success with it. Either it makes the dongle go into a bootloop, or crashes with some LIBUSB error

<details>

<summary>Runtime logs</summary>

```
pi@roxy:~ $ ./ReactCarplay-3.0.2-arm64.AppImage
/home/pi/.config/ReactCarplay
spawning carplay {
  dpi: 240,
  nightMode: 0,
  hand: 0,
  boxName: 'nodePlay',
  width: 1024,
  height: 600,
  fps: 30
}
directory found
device not connected
MESA-LOADER: failed to retrieve device information
MESA-LOADER: failed to retrieve device information
MESA-LOADER: failed to retrieve device information
MESA-LOADER: failed to open kms_swrast: /usr/lib/dri/kms_swrast_dri.so: cannot open shared object file: Permission denied (search paths /usr/lib/aarch64-linux-gnu/dri:\$${ORIGIN}/dri:/usr/lib/dri)
failed to load driver: kms_swrast
MESA-LOADER: failed to open swrast: /usr/lib/dri/swrast_dri.so: cannot open shared object file: Permission denied (search paths /usr/lib/aarch64-linux-gnu/dri:\$${ORIGIN}/dri:/usr/lib/dri)
failed to load swrast driver
carplay connection
carplay connection
[1250:0907/122433.504094:ERROR:gl_surface_presentation_helper.cc(260)] GetVSyncParametersIfAvailable() failed for 1 times!
sending dpi
unkown type:  18  with data::  <Buffer aa 55 aa 55 2e 00 00 00 12 00 00 00 ed ff ff ff>
parsing type:  8  for length:  4 <Buffer aa 55 aa 55 04 00 00 00 08 00 00 00 f7 ff ff ff>
parsing 8 <Buffer e9 03 00 00>
Carplay message:  undefined
test message i
unkown type:  38  with data::  <Buffer aa 55 aa 55 00 00 00 00 26 00 00 00 d9 ff ff ff>
FATAL ERROR:  LIBUSB_TRANSFER_ERROR
Aborted
```

</details>

rhysmorgan134 commented:
> Re `react-carplay`, I've never had any success with it. Either it makes the dongle go into a bootloop, or crashes with some LIBUSB error
> 
> Runtime logs

That’s interesting, what version firmware is the dongle? It’s a known issue that the 2023 version give the libusb error, perhaps Yonis rewriting has gotten around that but it is manifesting as an audio issue instead

steelbrain commented:
@rhysmorgan134 I do have 2023 model, the hardware version is `AW3035800` and the software version is `2023.05.29.1924`

The fix for `libusb` transfer error for my usecase at least was calling `reset` and then waiting 1-3 seconds and then opening the device again, since the node-usb impl of web usb is not that compliant (yet?)

steelbrain commented:
Okay so this is not related to the PCM player lib by itself, but all audio context situations, here's a patch of how I played audio through `@yume-chan/pcm-player` which still has the same issue

<details>

<summary>Patch for using Yume chan's player</summary>

```patch
diff --git a/examples/carplay-web-app-direct/package-lock.json b/examples/carplay-web-app-direct/package-lock.json
index 5a7e29d..df68049 100644
--- a/examples/carplay-web-app-direct/package-lock.json
+++ b/examples/carplay-web-app-direct/package-lock.json
@@ -15,6 +15,7 @@
         "@types/node": "^16.18.40",
         "@types/react": "^18.2.20",
         "@types/react-dom": "^18.2.7",
+        "@yume-chan/pcm-player": "^0.0.21",
         "buffer": "^6.0.3",
         "jmuxer": "^2.0.5",
         "node-carplay": "file:../../",
@@ -4345,6 +4346,14 @@
       "resolved": "https://registry.npmjs.org/@xtuc/long/-/long-4.2.2.tgz",
       "integrity": "sha512-NuHqBY1PB/D8xU6s/thBgOAiAP7HOYDQ32+BFZILJ8ivkUkAHQnWfn6WhL79Owj1qmUnoN/YPhktdIoucipkAQ=="
     },
+    "node_modules/@yume-chan/pcm-player": {
+      "version": "0.0.21",
+      "resolved": "https://registry.npmjs.org/@yume-chan/pcm-player/-/pcm-player-0.0.21.tgz",
+      "integrity": "sha512-2iPzfxML7lcHyIWGUYCxTxDJ59tcchSktrDXopcosQG9p8V6fhZ39NWSgPpcfPaSKaLO7mMXIRBR8ti7IdKdbw==",
+      "dependencies": {
+        "tslib": "^2.6.0"
+      }
+    },
     "node_modules/abab": {
       "version": "2.0.6",
       "resolved": "https://registry.npmjs.org/abab/-/abab-2.0.6.tgz",
diff --git a/examples/carplay-web-app-direct/package.json b/examples/carplay-web-app-direct/package.json
index e896631..51e6092 100644
--- a/examples/carplay-web-app-direct/package.json
+++ b/examples/carplay-web-app-direct/package.json
@@ -10,6 +10,7 @@
     "@types/node": "^16.18.40",
     "@types/react": "^18.2.20",
     "@types/react-dom": "^18.2.7",
+    "@yume-chan/pcm-player": "^0.0.21",
     "buffer": "^6.0.3",
     "jmuxer": "^2.0.5",
     "node-carplay": "file:../../",
diff --git a/examples/carplay-web-app-direct/src/useCarplayAudio.ts b/examples/carplay-web-app-direct/src/useCarplayAudio.ts
index 0903a0e..1c78a8b 100644
--- a/examples/carplay-web-app-direct/src/useCarplayAudio.ts
+++ b/examples/carplay-web-app-direct/src/useCarplayAudio.ts
@@ -7,7 +7,8 @@ import CarplayWeb, {
   AudioFormat,
   SendAudio,
 } from 'node-carplay/dist/web'
-import PCMPlayer from 'pcm-player'
+// import PCMPlayer from './vendor/pcm-player'
+import { Int16PcmPlayer } from '@yume-chan/pcm-player'
 
 const emptyFunc = () => {
   return {}
@@ -19,22 +20,24 @@ const defaultNavVolume = 0.5
 
 const useCarplayAudio = (carplay: CarplayWeb) => {
   const [mic, setMic] = useState<WebMicrophone | null>(null)
-  const [audioPlayers] = useState(new Map<AudioFormat, PCMPlayer>())
+  const [audioPlayers] = useState(new Map<AudioFormat, Int16PcmPlayer>())
 
   const getAudioPlayer = useCallback(
-    (format: AudioFormat): PCMPlayer => {
+    (format: AudioFormat) => {
       let player = audioPlayers.get(format)
       if (player) return player
-      player = new PCMPlayer({
-        channels: format.channel,
-        sampleRate: format.frequency,
-        inputCodec: 'Int16',
-        flushTime: 50,
-        onstatechange: emptyFunc,
-        onended: emptyFunc,
-      })
+      // player = new PCMPlayer({
+      //   channels: format.channel,
+      //   sampleRate: format.frequency,
+      //   inputCodec: 'Int16',
+      //   flushTime: 50,
+      //   onstatechange: emptyFunc,
+      //   onended: emptyFunc,
+      // })
+      player = new Int16PcmPlayer(format.frequency)
+      player.start()
       audioPlayers.set(format, player)
-      player.volume(defaultAudioVolume)
+      // player.volume(defaultAudioVolume)
       return player
     },
     [audioPlayers],
@@ -49,7 +52,7 @@ const useCarplayAudio = (carplay: CarplayWeb) => {
         } = audio
 
         const player = getAudioPlayer(format)
-        player.feed(audioData)
+        player.feed(new Int16Array(audioData))
       } else if (audio.command) {
         switch (audio.command) {
           case AudioCommand.AudioSiriStart:
@@ -61,13 +64,13 @@ const useCarplayAudio = (carplay: CarplayWeb) => {
             mic?.stop()
             break
           case AudioCommand.AudioNaviStart:
-            const navPlayer = getAudioPlayer(audio.format)
-            navPlayer.volume(defaultNavVolume)
+            // const navPlayer = getAudioPlayer(audio.format)
+            // navPlayer.volume(defaultNavVolume)
             break
           case AudioCommand.AudioMediaStart:
           case AudioCommand.AudioOutputStart:
-            const mediaPlayer = getAudioPlayer(audio.format)
-            mediaPlayer.volume(defaultAudioVolume)
+            // const mediaPlayer = getAudioPlayer(audio.format)
+            // mediaPlayer.volume(defaultAudioVolume)
             break
         }
       }
@@ -99,7 +102,8 @@ const useCarplayAudio = (carplay: CarplayWeb) => {
   // cleanup
   useEffect(() => {
     return () => {
-      audioPlayers.forEach(p => p.destroy())
+      // audioPlayers.forEach(p => p.destroy())
+      audioPlayers.forEach(p => p.stop())
       mic?.destroy()
     }
   }, [audioPlayers, mic])

```

</details>

gozmanyoni commented:
@steelbrain Yes I have tested various PCM players as well - it seems to happen across all of these to varying degrees.

However, all of  these solutions seem to be pretty heavy with creating buffers/arrays, and running flush intervals - I am trying out a different approach using a RingBuffer on this branch https://github.com/gozmanyoni/node-CarPlay/tree/pcm-player-shared

I used this - https://ringbuf-js.netlify.app/ which is a very handy and efficient ringbuffer implementation
There is a single SharedArrayBuffer between the main thread and the audio worklet (which runs on the audio thread) - the worklet keeps popping the required data size from the buffer and playing it - which i think is as efficient as it can get, with minimal GC.

There is still some static from what I can tell - but its different (hard to explain, or maybe I am imagining after listening to this repeatedly 🤣) - Do you mind verifying that my observation is correct ? :)

steelbrain commented:
Tried the branch, MUCH better than before. I'm listening to "Cowboy like me ~ Taylor Swift", and it's finally pretty clean on the first few seconds. You can hear some artefacts again around ~9 second mark but this is neat progress!

gozmanyoni commented:
@steelbrain @rhysmorgan134 I managed to get audio working smoothly.

I found an issue with how audio was parsed - we were passing the underlying arrayBuffer of the buffer object to the pcm-player, that is not bound to subarray(12) and contains the starting volume and audio type bytes.

These ofcourse should not be played :) and cause pops and clicks.

Together with the ringbuffer implementation audio is smooth and clear. I am still ironing out a few bugs but will release the pcm player as an npm package and open a PR fixing the examples

Instead we should create the PCM Int array like so, omitting the information bytes - `new Int16Array(data.buffer, 12)`

gozmanyoni commented:
@steelbrain - https://github.com/rhysmorgan134/node-CarPlay/pull/24 You can try using this branch here - works flawlessly on my machine :)

steelbrain commented:
Thank you! Moving the discussion to the referenced PR
